src/main/java/com/google/research/bleth/simulator/Simulation.java

```diff
@@ -1,9 +1,399 @@
+    private HashMap<String, Double> stats = new HashMap<>();
+
+    // A protected constructor used by the concrete simulation classes' constructors.
+    protected Simulation
+    (String id, int currentRound, int maxNumberOfRounds, IResolver resolver, double radius, Board realBoard,
+     List<Beacon> beacons, List<Observer> observers) {
+        this.id = id;
+        this.maxNumberOfRounds = maxNumberOfRounds;
+        this.board = realBoard;
+        this.resolver = resolver;
+        this.radius = radius;
+        this.beacons = ImmutableList.copyOf(beacons);
+        this.observers = ImmutableList.copyOf(observers);
+    }
+
+    /**
+     * Return the simulation real board.
+     * @return Board object representing the simulation's real state.
+     */
+    Board getBoard() {
         return board;
     }
+
+    /**
+     * Run entire simulation logic, including writing data to db.
+     */
+    public void run() {
+        if (currentRound == 0) {
+            writeRoundState(); // round 0 is the initial simulation state
+            currentRound++;
+        }
+        while (currentRound <= maxNumberOfRounds) {
+            moveAgents();
+            updateObserversAwaknessState();
+            beaconsToObservers();
+            observersToResolver();
+            resolverEstimate();
+            writeRoundState();
+            updateSimulationStats();
+            currentRound++;
+        }
+        writeSimulationStats();
+    }
+
+    /**
+     * Move all agents according to their movement strategies and update the real board.
+     */
+    void moveAgents() { }
+
+    /**
+     * Update all observers awakeness states according to their awakeness strategies.
+     */
+    void updateObserversAwaknessState() { }
+
+    /**
+     * Pass transmissions from beacons to observers while taking into consideration world-physics parameters,
+     * such as probability of transmission and distance between beacons and observers.
+     */
+    void beaconsToObservers() { }
+
+    /**
+     * Pass current-round information of transmission data from all observers to the simulation's resolver.
+     */
+    void observersToResolver() { }
+
+    /**
+     * Update resolver's estimated board.
+     */
+    void resolverEstimate() { }
+
+    /**
+     * Write current-round state of the simulation to db.
+     */
+    void writeRoundState() { }
+
+    /**
+     * Gather statistical data of the current round and update the aggregated simulation statistics based on all rounds.
+     */
+    abstract void updateSimulationStats();
+
+    /**
+     * Write final simulation statistical data to db.
+     */
+    void writeSimulationStats() { }
+
+    /**
+     * An abstract builder class designed to separate the construction of a simulation from its representation.
+     * Designated for the construction of a new simulation, based on parameters given by an end-user.
+     */
+    public static abstract class SimulationBuilder {
```

Comment:
Nit: since this is an inner class, common practice is to just call it "Builder" (and then always refer to it as "Simulation.Builder").